kv/kvserver/concurrency: TestConcurrencyManagerBasic failed

kv/kvserver/concurrency.TestConcurrencyManagerBasic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12159100?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12159100?buildTab=artifacts#/) on release-23.2 @ [4a76e70de36e2aaf051bb77117750eb3d54c0851](https://github.com/cockroachdb/cockroach/commits/4a76e70de36e2aaf051bb77117750eb3d54c0851):


```
        ----
        [5] sequence req2r: sequencing request
        [5] sequence req2r: acquiring latches
        [5] sequence req2r: scanning lock table for conflicting locks
        [5] sequence req2r: waiting in lock wait-queues
        [5] sequence req2r: lock wait-queue event: wait for (distinguished) txn 00000003 holding lock @ key ‹"c"› (queuedLockingRequests: 0, queuedReaders: 1)
        [5] sequence req2r: pushing after 0s for: liveness detection = true, deadlock detection = true, timeout enforcement = false, priority enforcement = false, wait policy error = false
        [5] sequence req2r: pushing timestamp of txn 00000003 above 10.000000000,1
        [5] sequence req2r: blocked on select in concurrency_test.(*cluster).PushTransaction
    datadriven.go:259: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11279/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/concurrency/concurrency_test_/concurrency_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks:130:
        sequence [1 args]
        <no input to command>
        ----
        [4] sequence req1r: dependency cycle detected 00000001->00000002->00000003->00000001
        [5] sequence req2r: dependency cycle detected 00000002->00000003->00000001->00000002
        [6] sequence req3r: sequencing request
        [6] sequence req3r: acquiring latches
        [6] sequence req3r: scanning lock table for conflicting locks
        [6] sequence req3r: waiting in lock wait-queues
        [6] sequence req3r: lock wait-queue event: wait for (distinguished) txn 00000001 holding lock @ key ‹"a"› (queuedLockingRequests: 0, queuedReaders: 1)
        [6] sequence req3r: pushing after 0s for: liveness detection = true, deadlock detection = true, timeout enforcement = false, priority enforcement = false, wait policy error = false
        [6] sequence req3r: pushing timestamp of txn 00000001 above 10.000000000,1
        [6] sequence req3r: blocked on select in concurrency_test.(*cluster).PushTransaction
        [6] sequence req3r: dependency cycle detected 00000003->00000001->00000002->00000003
    datadriven.go:259: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11279/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/concurrency/concurrency_test_/concurrency_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks:144:
        debug-lock-table [0 args]
        <no input to command>
        ----
        num=3
         lock: "a"
          holder: txn: 00000001-0000-0000-0000-000000000000 epoch: 0, iso: Serializable, ts: 10.000000000,1, info: unrepl [(str: Exclusive seq: 0)]
           waiting readers:
            req: 6, txn: 00000003-0000-0000-0000-000000000000
           distinguished req: 6
         lock: "b"
          holder: txn: 00000002-0000-0000-0000-000000000000 epoch: 0, iso: Serializable, ts: 10.000000000,1, info: unrepl [(str: Exclusive seq: 0)]
           waiting readers:
            req: 4, txn: 00000001-0000-0000-0000-000000000000
           distinguished req: 4
         lock: "c"
          holder: txn: 00000003-0000-0000-0000-000000000000 epoch: 0, iso: Serializable, ts: 10.000000000,1, info: unrepl [(str: Exclusive seq: 0)]
           waiting readers:
            req: 5, txn: 00000002-0000-0000-0000-000000000000
           distinguished req: 5
    concurrency_manager_test.go:1261: unknown goroutine state: sync.Mutex.Lock
        --- FAIL: TestConcurrencyManagerBasic/deadlocks/lock_ordering (0.17s)
=== RUN   TestConcurrencyManagerBasic/deadlocks
    --- FAIL: TestConcurrencyManagerBasic/deadlocks (0.17s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrencyManagerBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-32520

--- a/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
+++ b/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
@@ -1317,6 +1317,7 @@ var goroutineStalledStates = map[string]bool{
 	"waiting":   true,
 	"dead":      false,
 	"copystack": false,
+	"preempted": false,
 	"???":       false, // catch-all in runtime.goroutineheader
 
 	// runtime.goroutineheader may override these G statuses with a waitReason.
@@ -1344,13 +1345,22 @@ var goroutineStalledStates = map[string]bool{
 	// runtime itself. No request-level synchronization points use mutexes to
 	// wait for state transitions by other requests, so it is safe to ignore
 	// this state and wait for it to exit.
-	"semacquire":             false,
-	"sleep":                  false,
-	"sync.Cond.Wait":         true,
-	"timer goroutine (idle)": false,
+	"semacquire":     false,
+	"sleep":          false,
+	"sync.Cond.Wait": true,
+	// Similar to "semaquire" above, we mark the following three mutex states as
+	// non-stalled, assuming that they are transient states.
+	"sync.Mutex.Lock":        false,
+	"sync.RWMutex.RLock":     false,
+	"sync.RWMutex.Lock":      false,
 	"trace reader (blocked)": false,
 	"wait for GC cycle":      false,
 	"GC worker (idle)":       false,
+	"GC worker (active)":     false,
+	// "preempted" is already included above as part of gStatusStrings.
+	"debug call":          false,
+	"GC mark termination": false,
+	"stopping the world":  false,
 }
 
 // goroutineStatus returns a stack trace for each goroutine whose stack frame

--- a/pkg/testutils/lint/lint_test.go
+++ b/pkg/testutils/lint/lint_test.go
@@ -565,6 +565,8 @@ func TestLint(t *testing.T) {
 			":!util/syncutil/mutex_sync.go",
 			":!util/syncutil/mutex_sync_race.go",
 			":!testutils/lint/passes/deferunlockcheck/testdata/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync.go",
+			// Exception needed for goroutineStalledStates.
+			":!kv/kvserver/concurrency/concurrency_manager_test.go",
 		)
 		if err != nil {
 			t.Fatal(err)

